File: programmers/dijkstra/Q_12978_Baedal.java
package programmers.dijkstra;
import java.io.*;
import java.util.ArrayList;

public class Q_12978_Baedal {
    public static int solution(int N, int[][] road, int K) {
        int answer = 0;
        ArrayList<Edge>[] graph = makeList(N,road);
        boolean[] visited = new boolean[N + 1];
        dfs(graph,1,visited,K,0);

        return answer;
    }

    public static ArrayList[] makeList(int N, int[][] road){
        // 인접 리스트
        ArrayList<Edge>[] graph = new ArrayList[N+1];
        // 인접 리스트로 구성한 그래프에 ArrayList를 넣어주면서 초기화
        for(int i=0; i<graph.length; i++) {
            graph[i] = new ArrayList<Edge>();
        }

        for (int i = 0; i < road.length; i++) {
            int a = road[i][0];
            int b = road[i][1];
            int c = road[i][2];

            graph[a].add(new Edge(b,c));
        }

        return graph;
    }
    public static void dfs(ArrayList<Edge>[] graph, int cur, boolean[] visited, int K, int curDistance){
        visited[cur] = true;
        System.out.println(cur);

        for(Edge next : graph[cur]){
            if(next.to > 0 && !visited[next.to] && curDistance <=K){
                dfs(graph,next.to , visited, K, curDistance + next.distance);
            }
        }
//        for (int i = 1; i < graph.length; i++) {
//            if (graph[cur].to > 0 && !visited[i] && curDistance <= K) {
//                dfs(graph, i, visited, K,curDistance + graph[cur][i]);
//
//        }
    }

    public static class Edge {
        int to;
        int distance;

        public Edge(int to, int distance) {
            this.to = to;
            this.distance = distance;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

//        int N = 5;

//        int[][] road =  {
//                {1,2,1},
//                {2,3,3},
//                {5,2,2},
//                {1,4,2},
//                {5,3,1},
//                {5,4,2}
//        };
//        int K = 3;

        int N = 6;
        int[][] road = {
                {1,2,1},
                {1,3,2},
                {2,3,2},
                {3,4,3},
                {3,5,2},
                {3,5,3},
                {5,6,1}
        };
        int K = 4;

        solution(N, road, K);

        br.close();
        bw.close();
    }
}

File: programmers/dijkstra/Q_12978_Baedal.js
function solution(N, road, K) {
    var answer = 0;
    let graph = makeArray(N,road);
    graph[1][1] = 0;
    console.log(graph)
    let visited = new Array(N + 1);
    dfs(graph,1, visited, K,0);
    return answer;
}

function dfs(graph, cur, visited, K, curDistance){
    // 현재 마을 방문 처리
    visited[cur] = true;
    // console.log(graph.length);

    for (let i = 1; i < graph.length; i++) {
        // console.log(cur,i,graph[cur][i]);
        if(graph[cur][i] > 0 && !visited[i] && curDistance <= K ){
            dfs(graph, i, visited, curDistance + graph[cur][i]);
        }
    }
}

function makeArray(N, road){
    // 2차원 배열 만들기
    let graph = new Array(N+1);
    for (let i = 0; i < N+1 ; i++) {
        graph[i] = new Array(1);
    }

    console.log(graph)
    // 인접 행렬 만들기
    for (let i = 0; i <N ; i++) {
        let a = road[i][0];
        let b = road[i][1];
        let c = road[i][2];
        graph[a][b] = c;
    }

    console.log(graph[1][2])
    console.log(graph[2][3])
    console.log(graph[5][2])
    console.log(graph[1][4])
    console.log(graph[5][3])
    console.log(graph[5][4])
    return graph;
}

solution(5, [[1,2,1],[2,3,3],[5,2,2],[1,4,2],[5,3,1],[5,4,2]], 3)
// solution(6, [[1,2,1],[1,3,2],[2,3,2],[3,4,3],[3,5,2],[3,5,3],[5,6,1]], 4)
